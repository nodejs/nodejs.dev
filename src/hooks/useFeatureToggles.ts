import { useContext } from 'react';
import { FeatureToggleContext } from '../components/FeatureToggleProvider';

export const useFeatureToggles = () => {
  const featureFlags = useContext(FeatureToggleContext);

  return {
    has: (feature: string) => featureFlags.includes(feature),
    getAll: () => featureFlags,
  };
};
