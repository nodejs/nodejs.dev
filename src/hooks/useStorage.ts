import { useEffect } from 'react';

// Creates a permanent storage to be used between all the parts of the Application
// So it doesn't matter where this hook gets created it will be global.
// Note.: This could be moved within a Provider to move it to a React-isolated context-tree instead of a global runtime one.
export const appStorage = new Map<string, string>();

const syncWithBrowser = () => {
  if (typeof localStorage === 'object') {
    return {
      sendToBrowser: (k: string, v: string) => localStorage.setItem(k, v),
      getFromBrowser: (k: string) => localStorage.getItem(k) || undefined,
    };
  }

  return {
    sendToBrowser: () => undefined,
    getFromBrowser: () => undefined,
  };
};

export const useStorage = () => {
  const { sendToBrowser, getFromBrowser } = syncWithBrowser();

  const setItem = (key: string, value: unknown) =>
    appStorage.set(key, JSON.stringify(value));

  const getItem = (key: string) => {
    // Attempts to retrieve items from the App Storage, but if they don't exist yet
    // They might exist in the Browser Local Storage. This is useful as we retrieve things
    // From the Web Storage API only if they don't exist in the App Storage
    const storageItem = appStorage.get(key) || getFromBrowser(key);
    const parsedValue = storageItem ? JSON.parse(storageItem) : undefined;

    // If a said item doesn't exist in the App Storage we will create one
    // On the App Storage that will either be the value from the Web Storage if it exists
    // Or `undefined`. This is an interesting approach as it will ensure that all requested keys
    // at one point will exist in the Web Storage, and keeps things in sync.
    if (!appStorage.has(key)) {
      setItem(key, parsedValue);
    }

    return parsedValue;
  };

  const flushToStorage = () =>
    [...appStorage.entries()].forEach(([k, v]) => sendToBrowser(k, v));

  useEffect(() => {
    // Instead of flushing to the localStorage on every change, we flush once the Application
    // is being unloaded, so it allows us to update all changes at once instead of constantly
    // invoking the Web Storage APIs for every change we need, as the "authoritative storage" is the app itself
    window.addEventListener('unload', flushToStorage);

    return () => window.removeEventListener('unload', flushToStorage);
  }, []);

  return { getItem, setItem };
};
