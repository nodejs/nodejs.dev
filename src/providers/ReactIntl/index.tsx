import React from 'react';
import { IntlProvider } from 'react-intl';

interface Props {
  locale: string;
  messages: Record<string, string>;
}

const ReactIntlProvider = ({
  children,
  ...props
}: React.PropsWithChildren<Props>) => (
  <IntlProvider locale={props.locale} messages={props.messages}>
    {children}
  </IntlProvider>
);

export default ReactIntlProvider;
