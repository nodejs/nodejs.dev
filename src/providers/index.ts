export { default as ReactIntl } from './ReactIntl';
export * from './FeatureToggles';
