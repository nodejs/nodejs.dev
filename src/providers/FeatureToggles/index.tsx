import React, { createContext, useMemo } from 'react';
import { useStorage } from '../../hooks/useStorage';

type Props = { children?: React.ReactNode };

const FEATURE_FLAGS_STORAGE = 'node_featureFlags';

export const FeatureToggleContext = createContext<string[]>([]);

export const FeatureToggleProvider: React.FC<Props> = ({ children }) => {
  const { getItem } = useStorage();

  const featureFlags = useMemo(() => {
    const parsedItem = getItem(FEATURE_FLAGS_STORAGE) || [];

    // We want to guarantee the parsedItem is truthy and and Array
    if (parsedItem && Array.isArray(parsedItem)) {
      // Lastly we filter non-string objects as we want to avoid code injection such as XSS
      // from starting from our feature flags
      return [...parsedItem].filter(f => typeof f === 'string' && f.length);
    }

    return [];
  }, [getItem]);

  return (
    <FeatureToggleContext.Provider value={featureFlags}>
      {children}
    </FeatureToggleContext.Provider>
  );
};
