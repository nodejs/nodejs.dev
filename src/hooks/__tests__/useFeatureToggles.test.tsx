import React from 'react';
import { render } from '@testing-library/react';
import { useFeatureToggles } from '../useFeatureToggles';
import { appStorage } from '../useStorage';
import { FeatureToggleProvider } from '../../providers';

describe('useFeatureToggles', () => {
  const FeatureFlagRenderer = (): JSX.Element => {
    const featureFlags = useFeatureToggles();

    return (
      <>
        has:{featureFlags.has('feature-flag-mock') ? 'true' : 'false'},all:
        {featureFlags.getAll().toString()}
      </>
    );
  };

  const setItemMock = jest.fn();
  const getItemMock = jest.fn();

  beforeEach(() => {
    Storage.prototype.setItem = setItemMock;
    Storage.prototype.getItem = getItemMock;

    appStorage.clear();
  });

  afterEach(() => {
    setItemMock.mockRestore();
    getItemMock.mockRestore();
  });

  it('should contain feature flag', () => {
    getItemMock.mockReturnValue('["feature-flag-mock"]');

    const { container } = render(
      <FeatureToggleProvider>
        <FeatureFlagRenderer />
      </FeatureToggleProvider>
    );

    expect(getItemMock).toHaveBeenCalledWith('node_featureFlags');
    expect(container).toMatchSnapshot();
  });

  it('should not contain feature flag', () => {
    getItemMock.mockReturnValue('[]');

    const { container } = render(
      <FeatureToggleProvider>
        <FeatureFlagRenderer />
      </FeatureToggleProvider>
    );

    expect(getItemMock).toHaveBeenCalledWith('node_featureFlags');
    expect(container).toMatchSnapshot();
  });

  it('should work if localstorage item does not exist ', () => {
    getItemMock.mockReturnValue(null);

    const { container } = render(
      <FeatureToggleProvider>
        <FeatureFlagRenderer />
      </FeatureToggleProvider>
    );

    expect(getItemMock).toHaveBeenCalledWith('node_featureFlags');
    expect(container).toMatchSnapshot();
  });
});
