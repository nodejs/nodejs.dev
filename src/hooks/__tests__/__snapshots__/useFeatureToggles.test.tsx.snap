// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useFeatureToggles should contain feature flag 1`] = `
<div>
  has:
  true
  ,all:
  feature-flag-mock
</div>
`;

exports[`useFeatureToggles should not contain feature flag 1`] = `
<div>
  has:
  false
  ,all:
</div>
`;

exports[`useFeatureToggles should work if localstorage item does not exist  1`] = `
<div>
  has:
  false
  ,all:
</div>
`;
